### DISCUSSION TITLE
What is the correct way to update vcpkg as a submodule?

### DISCUSSION THREAD

**Discussion by crgnam** [microsoft/vcpkg#39605] [Q&A]
I keep vcpkg as a submodule in my project, located in `external/vcpkg/` (it is the only submodule in my project)

I needed to update a version of one of my dependencies, requiring I move-up the baseline and thus update the version of vcpkg.  In my root CMakeLists.txt, I have the following:

```
set(MY_VCPKG_TOOLCHAIN_FILE "${CMAKE_CURRENT_SOURCE_DIR}/external/vcpkg/scripts/buildsystems/vcpkg.cmake")

find_package(Git QUIET)
if(GIT_FOUND AND EXISTS "${CMAKE_CURRENT_SOURCE_DIR}/.git")
# Update submodules as needed
    option(GIT_SUBMODULE "Check submodules during build" ON)
    if(GIT_SUBMODULE)
        message(STATUS "Submodule update")
        execute_process(COMMAND ${GIT_EXECUTABLE} submodule update --init
                        WORKING_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR}
                        RESULT_VARIABLE GIT_SUBMOD_RESULT)
        if(NOT GIT_SUBMOD_RESULT EQUAL "0")
            message(FATAL_ERROR "git submodule update --init failed with ${GIT_SUBMOD_RESULT}.")
        endif()
    endif()
endif()

set(CMAKE_TOOLCHAIN_FILE "${MY_VCPKG_TOOLCHAIN_FILE}" CACHE FILEPATH "Vcpkg toolchain file" FORCE)
```

This appears to work fine and fetches the submodule when the project is first cloned, and then uses the vcpkg manifest to fetch all dependencies.  Great!  The trouble starts when I ever try to update the vcpkg version, this has been an issue in the past as well, but its now getting to be very frustrating as more people are working on the projection.

My first attempt was to update it using this:
```
cd external/vcpkg
git checkout afa12e7292fb47771e619675d7915645fe0adb1b
cd ../../
```

Re-running CMake and it couldn't find a single one of my dependencies saying that none of the versions existed.  I reset everything, then tried:

```
git submodule update --remote --merge
git add external/vcpkg/
```

This appeared to work but then it just hanged on:

```
waiting to take filesystem lock on <...>/vcpkg/.vcpkg-root
```

So I finally just completely deleted the directory `external/vcpkg` and NOW everything seemed to work correctly.  But that means anyone working with this project is also going to have to manually go delete `external/vcpkg` when they pull any of these changes.  Is there a better way to handle this?  I am working on just ditching including vcpkg as a submodule and requiring users to have their own copy, but it would be nice to have an optional submodule copy or something (if the user just wants to quickly build the application, set a `-DUSE_VCPKG_SUBMODULE` or something like that).  This seems overly annoying, though I'm also not familiar with submodules so that might just be a problem with how submodules work

**Comment by Micke3rd** [comment]
> I keep vcpkg as a submodule in my project

Sure.
the pessimistic approach: Your source code links to the desired compiled package versions. Who creates this packages, or with what, is unimportant. Only the commit of this file counts.
the optimistic approach: vcpkg builds the components during your build main app process

If I understand you correctly, you are looking for something in between. Since the effort is significantly higher - what additional value would this bring to your team?

**Reply by crgnam** [reply]
Just so I'm following, are you suggesting removing vcpkg as a submodule, and just requiring people have their own copy?  This would solve all of my problems, so perhaps that is just the simplest way to go.  I had adopted this at somepoint when it seemed recommended to include vcpkg as a submodule, though that appears to no longer be the case.

**Reply by Micke3rd** [reply]
> are you suggesting removing vcpkg as a submodule

no, especially if thats the tool you use to build your dependecies.
How to maintain a submodule in connection with the parent branch is a GIT topic, not something specific to vcpkg. That's why I suspected something else behind your question: you want to connect vcpkg as a submodule in such a way, that the same vcpkg branch version always creates the same version of the packages = your dependencies. My answer was related to the latter.
